File: T1 Q4 - Symmetric/3DES.cpp
#include <botan/block_cipher.h>
#include <botan/hex.h>
#include <botan/auto_rng.h>
#include <iostream>

// Function to print data as hexadecimal
void print_hex(const std::string &label, const std::vector<uint8_t> &data)
{
    std::cout << '\n'
              << label << ": " << Botan::hex_encode(data);
}

// Function to convert a string to a vector of bytes
std::vector<uint8_t> string_to_bytes(const std::string &str)
{
    return std::vector<uint8_t>(str.begin(), str.end());
}

// Function to convert a vector of bytes to a string
std::string bytes_to_string(const std::vector<uint8_t> &bytes)
{
    return std::string(bytes.begin(), bytes.end());
}

// Function to perform encryption
std::vector<uint8_t> encrypt(const std::vector<uint8_t> &key, const std::vector<uint8_t> &plaintext)
{
    std::vector<uint8_t> ciphertext;
    try
    {
        // Create a 3DES cipher object
        auto cipher = Botan::BlockCipher::create("TripleDES");
        if (!cipher)
        {
            throw std::runtime_error("Failed to create cipher.");
        }

        print_hex(cipher->name() + " ____plain text", plaintext);

        cipher->set_key(key);

        // Encrypt the block
        ciphertext = plaintext; // Copy plaintext to ciphertext vector
        cipher->encrypt(ciphertext);
        print_hex(cipher->name() + " encrypted text", ciphertext);
    }
    catch (const std::exception &e)
    {
        std::cerr << "Encryption error: " << e.what() << '\n';
    }
    return ciphertext;
}

// Function to perform decryption
std::vector<uint8_t> decrypt(const std::vector<uint8_t> &key, const std::vector<uint8_t> &ciphertext)
{
    std::vector<uint8_t> decryptedtext;
    try
    {
        // Create a 3DES cipher object
        auto cipher = Botan::BlockCipher::create("TripleDES");
        if (!cipher)
        {
            throw std::runtime_error("Failed to create cipher.");
        }

        cipher->set_key(key);

        // Decrypt the block
        decryptedtext = ciphertext; // Copy ciphertext to decryptedtext vector
        cipher->decrypt(decryptedtext);
        print_hex(cipher->name() + " decrypted text", decryptedtext);
    }
    catch (const std::exception &e)
    {
        std::cerr << "Decryption error: " << e.what() << '\n';
    }
    return decryptedtext;
}

int main()
{
    Botan::AutoSeeded_RNG rng;

    // Generate a random key for 3DES (192 bits = 24 bytes)
    Botan::secure_vector<uint8_t> secure_key = rng.random_vec(24);
    std::vector<uint8_t> key(secure_key.begin(), secure_key.end());
    std::cout << "Generated Key: " << Botan::hex_encode(key) << std::endl;

    // The human-readable message to be encrypted
    std::string message = "This message is to be encrypted using TripleDES";

    // Convert the message to a vector of bytes
    std::vector<uint8_t> plaintext = string_to_bytes(message);

    // Perform encryption
    std::vector<uint8_t> ciphertext = encrypt(key, plaintext);

    // Perform decryption
    std::vector<uint8_t> decryptedtext = decrypt(key, ciphertext);

    // Convert the decrypted text back to a string
    std::string decrypted_message = bytes_to_string(decryptedtext);

    // Verify decryption matches original message
    if (decrypted_message == message)
    {
        std::cout << "\n\nReceived message: " + message;
        std::cout << "\nDecryption successful, message matches.\n\n";
    }
    else
    {
        std::cout << "\nDecryption failed, message does not match.\n\n";
    }

    return 0;
}

File: T1 Q4 - Symmetric/3DES.py
import botan3 as bt
import os

# Function to print data as hexadecimal
def print_hex(label, data):
    hex_data = data.hex()
    print(f'{label}: {hex_data}')

# Function to pad data to be a multiple of the block size
def pad(data, block_size):
    padding_len = block_size - (len(data) % block_size)
    padding = bytes([padding_len] * padding_len)
    return data + padding

# Function to unpad data
def unpad(data):
    padding_len = data[-1]
    return data[:-padding_len]

# Function to convert a string to bytes
def string_to_bytes(s):
    return s.encode('utf-8')

# Function to convert bytes to a string
def bytes_to_string(b):
    return b.decode('utf-8')

# Function to perform encryption
def encrypt(key, plaintext):
    try:
        block_size = 8  # TripleDES block size in bytes
        cipher = bt.SymmetricCipher('TripleDES/CBC')
        cipher.set_key(key)
        iv = os.urandom(block_size)  # 8 bytes IV for TripleDES
        cipher.start(iv)
        padded_plaintext = pad(plaintext, block_size)
        ciphertext = cipher.finish(padded_plaintext)
        print_hex("TripleDES ____plain text", plaintext)
        print_hex("TripleDES encrypted text", ciphertext)
        return iv, ciphertext
    except Exception as e:
        print(f"Encryption error: {e}")
        return b'', b''

# Function to perform decryption
def decrypt(key, iv, ciphertext):
    try:
        block_size = 8  # TripleDES block size in bytes
        cipher = bt.SymmetricCipher('TripleDES/CBC', False)
        cipher.set_key(key)
        cipher.start(iv)
        plaintext = cipher.finish(ciphertext)
        unpadded_plaintext = unpad(plaintext)
        print_hex("TripleDES decrypted text", unpadded_plaintext)
        return unpadded_plaintext
    except Exception as e:
        print(f"Decryption error: {e}")
        return b''

def main():
    # Generate a random key for TripleDES (192 bits = 24 bytes)
    key = os.urandom(24)
    print(f"Generated Key: {key.hex()}\n")

    # The human-readable message to be encrypted
    message = "This message is to be encrypted using TripleDES"
    plaintext = string_to_bytes(message)

    # Perform encryption
    iv, ciphertext = encrypt(key, plaintext)

    # Perform decryption
    decryptedtext = decrypt(key, iv, ciphertext)

    # Convert the decrypted text back to a string
    decrypted_message = bytes_to_string(decryptedtext)

    # Verify decryption matches original message
    if decrypted_message == message:
        print(f"\nReceived message: {decrypted_message}")
        print("Decryption successful, message matches.\n")
    else:
        print("\nDecryption failed, message does not match.\n")

if __name__ == "__main__":
    main()
